libsolidity/codegen/ir/IRGenerationContext.cpp

```diff
@@ -153,10 +153,70 @@ string IRGenerationContext::trySuccessConditionVariable(Expression const& _expre
 	return "trySuccessCondition_" + to_string(_expression.id());
 }
 
+Arity IRGenerationContext::functionArity(FunctionDefinition const& _function)
+{
+	FunctionType const* functionType = TypeProvider::function(_function)->asCallableFunction(false);
+	solAssert(functionType, "");
+	return functionArity(*functionType);
+}
+
+Arity IRGenerationContext::functionArity(FunctionType const& _functionType)
+{
+	return {
+		TupleType(_functionType.parameterTypes()).sizeOnStack(),
+		TupleType(_functionType.returnParameterTypes()).sizeOnStack()
+	};
+}
+
+string IRGenerationContext::internalDispatchFunctionName(Arity const& _arity)
+{
+	return "dispatch_internal"
+		"_in_" + to_string(_arity.in) +
+		"_out_" + to_string(_arity.out);
+}
+
 string IRGenerationContext::internalDispatch(Arity const& _arity)
 {
-	string funName = "dispatch_internal_in_" + to_string(_arity.in) + "_out_" + to_string(_arity.out);
-	return m_functions.createFunction(funName, [&]() {
+	// UNIMPLEMENTED: Internal library calls via pointers are not implemented yet.
+	// We're not generating code for internal library functions here even though it's possible
+	// to call them via pointers. Right now such calls end up triggering the `default` case in
+	// the switch above.
+	set<FunctionDefinition const*> functions;
+	for (auto const& contract: mostDerivedContract().annotation().linearizedBaseContracts)
+		for (FunctionDefinition const* function: contract->definedFunctions())
+			if (
+				FunctionType const* functionType = TypeProvider::function(*function)->asCallableFunction(false);
+				!function->isConstructor() &&
+				TupleType(functionType->parameterTypes()).sizeOnStack() == _arity.in &&
+				TupleType(functionType->returnParameterTypes()).sizeOnStack() == _arity.out
+			)
+			{
+				functions.insert(function);
+				enqueueFunctionForCodeGeneration(*function);
+			}
+
+	return internalDispatch(functions);
+}
+
+string IRGenerationContext::internalDispatch(Arity const& _arity, set<FunctionDefinition const*> const& _functions)
+{
+	vector<map<string, string>> cases;
+	for (auto const* function: _functions)
+	{
+		solAssert(function, "");
+		solAssert(functionArity(*function) == _arity, "A single dispatch function can only handle functions of one arity");
+		solAssert(!function->isConstructor(), "");
+		// 0 is reserved for uninitialized function pointers
+		solAssert(function->id() != 0, "Unexpected function ID: 0");
+
+		cases.emplace_back(map<string, string>{
+			{"funID", to_string(function->id())},
+			{"name", functionName(*function)}
+		});
+	}
+
+	string funName = internalDispatchFunctionName(_arity);
```

Comment:
Since the function is only generated once, it is very important that the `_functions` argument is correct. If the `internalDispatch` function is called again, it will not re-generate the function. Is there any way to ensure that?

Reply:
Sure. I'll add an assert that the dispatch function is not already in the function collector.

Reply:
Done.